File: Java/remove-duplicates-from-sorted-list.java
//  Remove Duplicates from Sorted List

/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */

// Time O(N), Space O(1)
class Solution {
    public ListNode deleteDuplicates(ListNode head) {
        if (head==null)
            return null;
        ListNode pre = head, cur = head.next;
        while (cur != null) {
            if (cur.val != pre.val) {
                pre.next = cur;
                pre = cur;
            }
            cur = cur.next;
        }
        pre.next = cur;
        return head;
    }
}

File: leetcode/remove-duplicates-from-sorted-list.cc
//  Remove Duplicates from Sorted List

/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode* deleteDuplicates(ListNode* head) {
        unordered_map<int, int> map;
        for (ListNode *p = head; p; p = p->next)
            map[p->val] ++;
        ListNode *dummy = new ListNode(-1);
        dummy->next = head;
        ListNode *cur = head, *pre = dummy;
        while (cur) {
            if (map[cur->val] > 1) {
                pre->next = cur->next;
                cur = pre->next;
            }
            else {
                pre = cur;
                cur = cur->next;
            }
        }
        return dummy->next;
    }
};
